File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.py
from collections import defaultdict


def find(tree, x):
    if tree[x] == x:
        return x
    
    tree[x] = find(tree, tree[x])
    
    return tree[x]


def union(tree, rank, x, y):
    x_root = find(tree, x)
    y_root = find(tree, y)
    
    if x_root == y_root:
        return
    
    if rank[x_root] < rank[y_root]:
        tree[x_root] = y_root
    elif rank[x_root] > rank[y_root]:
        tree[y_root] = x_root
    else:
        tree[y_root] = x_root
        rank[x_root] += 1


def solution(n, costs):
    answer = 0
    
    costs.sort(key = lambda x : x[2])
    
    tree = [i for i in range(n)]
    rank = [0 for _ in range(n)]
    
    for u, v, c in costs:
        if find(tree, u) != find(tree, v):
            union(tree, rank, u, v)
            answer += c
    
    return answer

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.kt
class Solution {
    fun find(tree: IntArray, x: Int): Int {
        if (tree[x] == x) return x
        
        tree[x] = find(tree, tree[x])
        
        return tree[x]
    }
    
    fun union(tree: IntArray, rank: IntArray, x: Int, y: Int): Unit {
        val xroot = find(tree, x)
        val yroot = find(tree, y)
        
        if (rank[xroot] < rank[yroot]) {
            tree[xroot] = yroot
        } else if (rank[xroot] > rank[yroot]) {
            tree[yroot] = xroot
        } else {
            tree[yroot] = xroot
            rank[xroot]++
        }
    }
    
    fun solution(n: Int, costs: Array<IntArray>): Int {
        costs.sortBy({it[2]})
        
        val tree = IntArray(n) {it}
        val rank = IntArray(n) {0}
        
        var minCost = 0
        var edges = 0
        
        for (edge in costs) {
            if (edges == n - 1) break
            
            val x = find(tree, edge[0])
            val y = find(tree, edge[1])
            
            if (x != y) {
                union(tree, rank, x, y)
                minCost += edge[2]
                edges++
            }
        }
        
        return minCost
    }
}